src/Einsum.jl

```diff
@@ -27,336 +28,404 @@ macro einsum_checkinbounds(ex)
-                         arr_store::Vector{Symbol},
-                         dim_store::Vector{Expr})
-    return idx_store, arr_store, dim_store
+function extractindices!(expr::Number,
+                         array_names::Vector{Symbol},
+                         index_names::Vector{Symbol},
+                         axis_expressions::Vector{Expr})
+    return array_names, index_names, axis_expressions
 end
 
 
-function extractindices!(ex::Expr,
-                         idx_store::Vector{Symbol},
-                         arr_store::Vector{Symbol},
-                         dim_store::Vector{Expr})
-    if Meta.isexpr(ex, :ref) # e.g. A[i,j,k]
-        arrname = ex.args[1]
-        push!(arr_store, arrname)
+function extractindices!(expr::Expr,
+                         array_names::Vector{Symbol},
+                         index_names::Vector{Symbol},
+                         axis_expressions::Vector{Expr})
+    if Meta.isexpr(expr, :ref) # e.g. A[i,j,k]
+        array_name = expr.args[1]
+        push!(array_names, array_name)
 
-        # ex.args[2:end] are indices (e.g. [i,j,k])
-        for (pos, idx) in enumerate(ex.args[2:end])
-            extractindex!(idx, arrname, pos, idx_store, arr_store, dim_store)
+        # expr.args[2:end] are indices (e.g. [i,j,k])
+        for (dimension, index_expr) in enumerate(expr.args[2:end])
+            extractindex!(index_expr, array_name, dimension,
+                          array_names, index_names, axis_expressions)
         end
-    elseif Meta.isexpr(ex, :call)
+    elseif Meta.isexpr(expr, :call)
         # e.g. 2*A[i,j], transpose(A[i,j]), or A[i] + B[j], so
-        # ex.args[2:end] are the individual tensor expressions (e.g. [A[i], B[j]])
-        for arg in ex.args[2:end]
-            extractindices!(arg, idx_store, arr_store, dim_store)
+        # expr.args[2:end] recursively contain the individual tensor
+        # expressions (e.g. [A[i], B[j]])
+        for arg in expr.args[2:end]
+            extractindices!(arg, array_names, index_names, axis_expressions)
         end
     else
-        throw(ArgumentError("Invalid expression head: `:$(ex.head)`"))
+        throw(ArgumentError("Invalid expression head: `:$(expr.head)`"))
     end
 
-    return idx_store, arr_store, dim_store
+    return return array_names, index_names, axis_expressions
 end
 
-function extractindex!(ex::Symbol, arrname, position,
-                       idx_store, arr_store, dim_store)
-    push!(idx_store, ex)
-    push!(dim_store, :(size($arrname, $position)))
-    return idx_store, arr_store, dim_store
+function extractindex!(expr::Symbol, array_name::Symbol, dimension::Int,
+                       array_names::Vector{Symbol},
+                       index_names::Vector{Symbol},
+                       axis_expressions::Vector{Expr})
+    push!(index_names, expr)
+    push!(axis_expressions, :(size($array_name, $dimension)))
+    return array_names, index_names, axis_expressions
```

Comment:
The return value of `extractindex!` (singular) isn't used, perhaps clearer to remove it: 
```suggestion
    return nothing
```